pkg/reconciler/dependencybuild/dependencybuild.go

```diff
@@ -166,61 +195,66 @@ func (r *ReconcileDependencyBuild) handleStateSubmitBuild(ctx context.Context, d
 	}
 	//now we submit the build
 	if err := r.client.Create(ctx, &tr); err != nil {
+		if errors.IsAlreadyExists(err) {
+			log.V(4).Info("handleStateBuilding: taskrun %s:%s already exists, not retrying", tr.Namespace, tr.Name)
+			return reconcile.Result{}, nil
+		}
 		r.eventRecorder.Eventf(db, v1.EventTypeWarning, "TaskRunCreationFailed", "The DependencyBuild %s/%s failed to create its build pipeline run", db.Namespace, db.Name)
 		return reconcile.Result{}, err
 	}
+
 	return reconcile.Result{}, nil
 }
 
-func (r *ReconcileDependencyBuild) handleStateBuilding(ctx context.Context, depId string, db *v1alpha1.DependencyBuild) (reconcile.Result, error) {
-	//make sure we still have a linked pr
-	list := &pipelinev1beta1.TaskRunList{}
-	lbls := map[string]string{
-		artifactbuild.DependencyBuildIdLabel: depId,
-	}
-	listOpts := &client.ListOptions{
-		Namespace:     db.Namespace,
-		LabelSelector: labels.SelectorFromSet(lbls),
-	}
-	if err := r.nonCachingClient.List(ctx, list, listOpts); err != nil {
-		return reconcile.Result{}, err
-	}
-	if len(list.Items) == 0 {
-		//no linked pr, this seems to happen randomly where the PR does not show up
-		//just return and next reconcile it is there
-		//TODO: Why is this happening? caching?
-		r.eventRecorder.Eventf(db, v1.EventTypeWarning, "NoTaskRun", "The DependencyBuild %s/%s did not have any TaskRuns", db.Namespace, db.Name)
-		return reconcile.Result{}, nil
-	}
-	var pr *pipelinev1beta1.TaskRun
-	//look for the most recent one, there could be multiple builds if earlier recipes failed
-	for _, current := range list.Items {
-		if pr == nil || pr.CreationTimestamp.Before(&current.CreationTimestamp) {
-			pr = &current
+func (r *ReconcileDependencyBuild) handleTaskRunReceived(ctx context.Context, tr *pipelinev1beta1.TaskRun) (reconcile.Result, error) {
+	if tr.Status.CompletionTime != nil {
+		// get db
+		ownerRefs := tr.GetOwnerReferences()
+		if ownerRefs == nil || len(ownerRefs) == 0 {
+			msg := "taskrun missing onwerrefs %s:%s"
+			r.eventRecorder.Eventf(tr, v1.EventTypeWarning, msg, tr.Namespace, tr.Name)
+			log.Info(msg, tr.Namespace, tr.Name)
+			return reconcile.Result{}, nil
 		}
-	}
-	if pr.Status.CompletionTime != nil {
-		if pr.Name == db.Status.LastCompletedBuildTaskRun {
+		if len(ownerRefs) > 1 {
+			// workaround for event/logging methods that can only take string args
+			count := fmt.Sprintf("%d", len(ownerRefs))
+			msg := "taskrun %s:%s has %s ownerrefs but only using the first"
+			r.eventRecorder.Eventf(tr, v1.EventTypeWarning, msg, tr.Namespace, tr.Name, count)
+			log.Info(msg, tr.Namespace, tr.Name, count)
+		}
+		ownerRef := ownerRefs[0]
+		key := types.NamespacedName{Namespace: tr.Namespace, Name: ownerRef.Name}
+		db := v1alpha1.DependencyBuild{}
+		err := r.client.Get(ctx, key, &db)
+		if err != nil {
+			msg := "get for taskrun %s:%s owning db %s:%s yielded error %s (GGM DO WE STILL NEED RAW CLIENT!?!?!?)"
```

Comment:
realized my prior analysis may have not distinguished from an issue here

I've reset my test system though, so I left this here, will add something similar in the artifactbuild flow, and scan again next time